File: frontend/src/test/csv.test.js
import { describe, it, expect } from 'vitest'
import { csvCell, csvRows } from '../utils/csv.js'

/**
 * CSV hücresi kaçışı — arayüz dışa aktarımlarının ortak kuralı.
 *
 * <p>Backend'deki {@code com.sitemonitor.util.Csv} ile AYNI davranışı bekler; iki taraf ayrışırsa
 * aynı veri iki dosyada farklı kaçışlanır. 2026-08-23 denetiminde arayüzdeki dört dışa aktarımın
 * hiçbirinde formül nötrlemesi olmadığı, üçünde de CR'nin kaçırıldığı bulundu.
 */
describe('csvCell', () => {
  it('formül karakteriyle BAŞLAYAN hücreyi metne sabitler', () => {
    expect(csvCell("=cmd|'/c calc'!A1")).toBe("'=cmd|'/c calc'!A1")
    expect(csvCell('+1+1')).toBe("'+1+1")
    expect(csvCell('-2+3')).toBe("'-2+3")
    expect(csvCell('@SUM(A1)')).toBe("'@SUM(A1)")
    expect(csvCell('\tsekme')).toBe("'\tsekme")
  })

  it('formül karakteri ORTADA ise dokunmaz', () => {
    expect(csvCell('a=b')).toBe('a=b')
    expect(csvCell('mail@akbank.com')).toBe('mail@akbank.com')
  })

  it('ayraç/tırnak/satır sonu içeren değeri tırnaklar, tırnakları ikiler', () => {
    expect(csvCell('a,b')).toBe('"a,b"')
    expect(csvCell('a;b')).toBe('"a;b"')
    expect(csvCell('de"mek')).toBe('"de""mek"')
    expect(csvCell('iki\nsatır')).toBe('"iki\nsatır"')
  })

  it('tek başına CR de tırnaklanır — satırlar CRLF ile birleşiyor', () => {
    expect(csvCell('a\rb')).toBe('"a\rb"')
  })

  it('nötrleme ve tırnaklama birlikte uygulanır', () => {
    expect(csvCell('=1,2')).toBe('"\'=1,2"')
  })

  it('null/undefined/boş → boş hücre', () => {
    expect(csvCell(null)).toBe('')
    expect(csvCell(undefined)).toBe('')
    expect(csvCell('')).toBe('')
  })

  it('sayı ve boolean da kabul edilir; negatif sayı formül gibi başladığı için nötrlenir', () => {
    expect(csvCell(443)).toBe('443')
    expect(csvCell(true)).toBe('true')
    expect(csvCell(-5)).toBe("'-5")
  })
})

describe('csvRows', () => {
  it('satırları CRLF ile birleştirir ve her hücreyi kaçışlar', () => {
    expect(csvRows([['a', 'b,c'], ['=x', 2]])).toBe('a,"b,c"\r\n\'=x,2')
  })

  it('boş liste boş gövde döner', () => {
    expect(csvRows([])).toBe('')
  })
})

File: backend/src/test/java/com/sitemonitor/util/CsvTest.java
package com.sitemonitor.util;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * CSV hücresi kaçışı — dışa aktarımların ortak kuralı.
 *
 * <p>Kural projede vardı ama tek bir dışa aktarımda: denetim kaydı CSV'si, kontrol geçmişi CSV'si
 * ve arayüzdeki dört dışa aktarım kendi kaçışını yazmıştı, hiçbirinde FORMÜL nötrlemesi yoktu
 * (2026-08-23 denetimi). Kural artık tek yerde ve burada kilitli.
 */
class CsvTest {

    @Test
    @DisplayName("Formül karakteriyle BAŞLAYAN hücre metne sabitlenir (Excel/LibreOffice çalıştırmasın)")
    void neutralisesFormulaPrefixes() {
        assertThat(Csv.cell("=cmd|'/c calc'!A1")).isEqualTo("'=cmd|'/c calc'!A1");
        assertThat(Csv.cell("+1+1")).isEqualTo("'+1+1");
        assertThat(Csv.cell("-2+3")).isEqualTo("'-2+3");
        assertThat(Csv.cell("@SUM(A1)")).isEqualTo("'@SUM(A1)");
        assertThat(Csv.cell("\tsekme")).isEqualTo("'\tsekme");
    }

    @Test
    @DisplayName("Formül karakteri ORTADA ise dokunulmaz — normal metin bozulmasın")
    void leavesInnerCharactersAlone() {
        assertThat(Csv.cell("a=b")).isEqualTo("a=b");
        assertThat(Csv.cell("1-2")).isEqualTo("1-2");
        assertThat(Csv.cell("mail@akbank.com")).isEqualTo("mail@akbank.com");
    }

    @Test
    @DisplayName("Ayraç, tırnak ve satır sonu içeren değer tırnaklanır; tırnaklar ikilenir")
    void quotesWhenNeeded() {
        assertThat(Csv.cell("a,b")).isEqualTo("\"a,b\"");
        assertThat(Csv.cell("a;b")).isEqualTo("\"a;b\"");
        assertThat(Csv.cell("de\"mek")).isEqualTo("\"de\"\"mek\"");
        assertThat(Csv.cell("iki\nsatır")).isEqualTo("\"iki\nsatır\"");
    }

    @Test
    @DisplayName("Tek başına CR de tırnaklanır — satırlar CRLF ile birleşiyor, yoksa satır bölünür")
    void carriageReturnIsQuoted() {
        assertThat(Csv.cell("a\rb")).isEqualTo("\"a\rb\"");
    }

    @Test
    @DisplayName("CR ile BAŞLAYAN değer hem nötrlenir hem tırnaklanır")
    void leadingCarriageReturnIsBothNeutralisedAndQuoted() {
        assertThat(Csv.cell("\rgizli")).isEqualTo("\"'\rgizli\"");
    }

    @Test
    @DisplayName("Formül nötrlemesi tırnaklamayı BOZMAZ (ikisi birlikte uygulanır)")
    void neutralisationAndQuotingCompose() {
        assertThat(Csv.cell("=1,2")).isEqualTo("\"'=1,2\"");
    }

    @Test
    @DisplayName("null ve boş değer boş hücre olur")
    void nullAndEmpty() {
        assertThat(Csv.cell(null)).isEmpty();
        assertThat(Csv.cell("")).isEmpty();
    }

    @Test
    @DisplayName("Metin olmayan değerler de kabul edilir (sayı/boolean)")
    void nonStringValues() {
        assertThat(Csv.cell(443)).isEqualTo("443");
        assertThat(Csv.cell(Boolean.TRUE)).isEqualTo("true");
        assertThat(Csv.cell(-5)).isEqualTo("'-5");   // negatif sayı da formül gibi başlar
    }
}
